ddtrace/appsec/iast/taint_sinks/_base.py

```diff
@@ -1,28 +1,88 @@
 from typing import TYPE_CHECKING
 
 from ddtrace import tracer
+from ddtrace.appsec.iast import oce
+from ddtrace.appsec.iast.overhead_control_engine import Operation
+from ddtrace.appsec.iast.reporter import Evidence
+from ddtrace.appsec.iast.reporter import IastSpanReporter
+from ddtrace.appsec.iast.reporter import Location
+from ddtrace.appsec.iast.reporter import Vulnerability
+from ddtrace.appsec.iast.stacktrace import get_info_frame
+from ddtrace.constants import IAST_CONTEXT_KEY
+from ddtrace.internal import _context
 from ddtrace.internal.logger import get_logger
+from ddtrace.vendor.wrapt import wrap_function_wrapper
 
 
 if TYPE_CHECKING:  # pragma: no cover
     from typing import Any
     from typing import Callable
+    from typing import Text
 
 log = get_logger(__name__)
 
 
-def inject_span(func):
-    # type: (Callable) -> Callable
-    """Get the current root Span and attach it to the wrapped function. We need the span to report the vulnerability
-    and update the context with the report information.
-    """
+class VulnerabilityBase(Operation):
+    vulnerability_type = ""
+    evidence_type = ""
 
-    def wrapper(wrapped, instance, args, kwargs):
-        # type: (Callable, Any, Any, Any) -> Any
-        span = tracer.current_root_span()
-        if span:
-            return func(wrapped, span, instance, args, kwargs)
-        log.warning("No root span in the current execution. Skipping IAST Taint sink.")
-        return wrapped(*args, **kwargs)
+    @classmethod
+    def wrap(cls, func):
+        # type: (Callable) -> Callable
+        def wrapper(wrapped, instance, args, kwargs):
+            # type: (Callable, Any, Any, Any) -> Any
+            """Get the current root Span and attach it to the wrapped function. We need the span to report the vulnerability
+            and update the context with the report information.
+            """
+            if oce.request_has_quota and cls.has_quota():
+                return func(wrapped, instance, args, kwargs)
+            else:
+                log.debug("IAST: no vulnerability quota to analyze more sink points")
+            return wrapped(*args, **kwargs)
 
-    return wrapper
+        return wrapper
+
+    @classmethod
+    def report(cls, evidence_value=""):
+        # type: (Text) -> None
+        """Build a IastSpanReporter instance to report it in the `AppSecIastSpanProcessor` as a string JSON
+
+        TODO: check deduplications if DD_IAST_DEDUPLICATION_ENABLED is true
+        """
+        if cls.has_quota():
+
+            span = tracer.current_root_span()
+            if not span:
+                log.debug("No root span in the current execution. Skipping IAST taint sink.")
+                return None
```

Comment:
I'd move this check outside of `has_quota` to lower the pressure over the threading lock